wgpu/src/lib.rs

```diff
@@ -2188,6 +2191,20 @@ impl Drop for Device {
     }
 }
 
+/// Error when parsing shader source that isn't wgsl, thus converting to a naga module beforehand.
+#[derive(Debug)]
+pub enum CreateShaderModuleError {
```

Comment:
I just want to note that this is the first time naga becomes a public dependency.

Reply:
Would it make sense to do a `Box<dyn Error + Send + Sync + 'static>` here, to avoid that?

Reply:
I'm not sure how to impl Error on this error since glsl's parsers return multiple errors. For example what should source() return ?

Reply:
Perhaps `naga` could provide an error type which wraps multiple errors where they can occur, so that the `source()` of this error could be that error type. Then, if this enum (or its variants' fields) are hidden in a struct with private fields (or, I suppose, a `Box<dyn Error + ...`), the `naga` types become invisible except when the caller chooses to attempt downcasting to them.

Reply:
P.S. It would be good if, whatever the type structure turns out to be, the result of printing the error (chain) generically is a readable formatted report of the kind naga can produce. (I mentioned this already in #2545 but it seems worth reiterating as a design goal here.)

Reply:
The root difficulty here is that Rust's `Error` trait simply is not designed to represent a list of errors like the kind a compiler front end would produce. This is why questions like "What should `source()` return?" have no good answer. Also, note that `naga::front::glsl::Parser::parse` returns `Result<Module, Vec<Error>>` - and `Vec<Error>` does not implement `Error`. So Naga's API is stretching Rust's `Result` paradigm a bit too far.

In the WebGPU API for content JavaScript, [`createShaderModule`](https://www.w3.org/TR/webgpu/#dom-gpudevice-createshadermodule) always succeeds, but the `GPUShaderModule` it returns may be marked "invalid". Valid or not, one can always call [`compilationInfo`](https://www.w3.org/TR/webgpu/#dom-gpushadermodule-compilationinfo) on a `GPUShaderModule` to retrieve a [`GPUCompilationInfo`](https://www.w3.org/TR/webgpu/#shader-module-compilation-information) structure, which has a list of error messages. Since the `GPUCompilationInfo` struct is available even when compilation succeeds, it can also carry warning messages.

Perhaps the thorough approach here would be to make `wgpu` match WebGPU more closely:
- add a `CompilationInfo` struct to `wgpu`, modeled after WebGPU's `GPUCompilationInfo`
- add an `Option<Box<CompilationInfo>>` to `wgpu::ShaderModule`
- add a method to `ShaderModule` to retrieve it
- add a method to `wgpu_core::hub::Global` akin to [`Global::create_buffer_error`](https://github.com/gfx-rs/wgpu/blob/f4c01052ef709e010f4e228d7d7d668f0166b3f5/wgpu-core/src/device/mod.rs#L3289-L3324) for allocating a dead-on-arrival shader module id, so that you can actually return a `ShaderModule` even when the Naga functions return an error.

-

Reply:
I agree that Naga's error messages with spans and labels are really nice - nicer than WebGPU's `GPUCompilationMessage`, for sure. I guess `wgpu`'s `CompilationInfo` should follow Naga, and then it'd be Firefox's problem to convert that into what WebGPU requires.

Reply:
I don't think this is a good first issue anymore :D I'll be closing the PR since it seems like there's too much at stake here and I'm not comfortable breaking new ground on wgpu. I think you summed it up well though.